### ISSUE TITLE
ISA circuit support for latest Runtime

### ISSUE THREAD

**Issue by woodsp-ibm** [qiskit-community/qiskit-algorithms#164]
https://docs.quantum.ibm.com/announcements/product-updates/2024-02-14-qiskit-runtime-primitives-update

> **beginning 1 March 2024**, Qiskit Runtime will require that circuits and observables are transformed to use only instructions supported by the system (referred to as instruction set architecture (ISA) circuits and observables) before being submitted to the primitives

This affects the existing V1 primitives, as supported here, as well as the newly introduced V2 primitives which are yet to be supported, see #136

While some algorithms, like VQE, can already be passed a transpiled ansatz and adjusted operator(s),  others build out the circuit internally to run so to continue supporting these they require change.

Any solution to this, to update algorithms here, needs further discussion etc

**Comment by woodsp-ibm** [comment]
I am just going to add these thoughts. I do not have a specific solution and  the thought I had about what might be a compatible path for users so existing code could continue to work is not without problems. Anyway here is the thought....

And as far as any fixes in general I did think perhaps a TranspileEngine or some common class that could be called upon to transpile by any of the algos. I was thinking a global instance being created of that, like the random number generator in utils, that an algo could just use. Nothing would be needed to be passed into any algo so the code would work as is. The engine could have some settings for skip transpilation, backend, opt level etc. And maybe even allow you to give it a passmanager etc. Evidently all algos would need to call this global instance to transpile their circuits, and if skipping it would just return the same deal. Algos could even add a new parameter for a transpile engine so you could do things differently/independently and not have to worry about coordinating on the global instance. Custom versions of the engine leveraging alternative transpilation mechanisms etc could be done allowing flexibility for the future. I was imagining an algo would use this to transpile the circuits it needs to run and then call the Estimator/Sampler primitive using the transpiled version.

Now this thought is not without problems. Lets take say QSVC in Qiskit Machine Learning. It knows the circuit, which is passed to it, and if it did things directly it would transpile (any resultant from) that circuit once and then use it. However it uses ComputeUncompute which composes the circuit to a inverse of itself and it's that which needs transpiling - passing a transpiled circuit to ComputeUncompute does not work (its still an ISA circuit to the runtime). Now ComputeUncompute could transpile at it's level for each fidelity computation, which would work, but it would be very inefficient since its the same circuit where only QSVC knows this. V1 primitives did do things like id circuits to avoid this so that's possible here or do some managed dictionary of them in UncomputeCompute where a circuit is added to dictionary and that dict key is what is used to do the fidelity not the circuit itself. Of course that diverges the interface away from how other primitives look (the Fidelities being consider to look/function like primitives). I imagine Gradients here may have similar considerations.

The above problem stated succinctly is if there is a separation between the algorithm that owns/manages the original circuits which then passes it/them to another which does changes to the circuit, how to manage the transpilation to just when needed - the original algo may well know the circuits are the same being passed each time, the receiver just sees circuits each time and has no idea - without resort to id'ing them as did the V1 primitives.

**Comment by BrunoRosendo** [comment]
I'm thinking about this workaround: getting a local copy of qiskit-algorithms and change these 2 lines (mixer and initial_state) to the transpiled versions of the circuits

https://github.com/qiskit-community/qiskit-algorithms/blob/3859d63ad31847e6f5820030a079e3c085e54180/qiskit_algorithms/minimum_eigensolvers/qaoa.py#L124

Would something like this work?

```python
self.mixer = qiskit.compiler.transpile(mixer, backend=backend)
self.initial_state = qiskit.compiler.transpile(initial_state, backend=backend)
```

**Comment by BrunoRosendo** [comment]
Hi. I am commenting again to ask if there's a fix coming anytime soon or if a workaround with a local copy of the library is the only way to go about this

**Comment by woodsp-ibm** [comment]
I will also note that for IBM Runtime that the ISA circuit is just one aspect. As of Runtime 0.23 the V1 primitives, which is all the algorithms support, are deprecated and would be removed in the not too distant future. If this QAOA code is you want/need to use then doing a local workaround for just your needs, and not waiting for some more general change to come, seems like the practical/sensible choice - a) to deal with ISA and b) to use V2 primitive instead of V1.

**Comment by BrunoRosendo** [comment]
I went with @woodsp-ibm's suggestion and implemented a solution for my use case in a local copy of mine. I created a fork, in case anybody is in the same situation and needs a working version immediately: https://github.com/BrunoRosendo/qiskit-algorithms

However, note that this was only tested for MinimumEigenOptimizer. This should work for other solvers if the operator passed is a SparsePauliOp. This is the case because I use the `apply_layout` method, which is not present in the abstract BaseOperator accepted by QAOA/SamplingVQE.

As for the solution, it is surprisingly simple, being composed by the ansatz transpilation and applying its layout to the operator.

**Comment by woodsp-ibm** [comment]
> As for the solution, it is surprisingly simple,

Yes, for some algos, where the circuit passed is what ends up being given to primitive that works out simple enough. Some algos here, e.g. grover are more complicated. Same with ComputeUncompute fidelty which appends the inverse of the circuit to whatever is given - the result gets rejected as not ISA even with the input being ISA. Hence the discussion here being to have some transpile engine algos can use - VQE could support no transpilation too allowing one to pass their own ISA circuit and suitable operators as you did. This is not the case for everything though so a solution that spans the codebase here is rather more work.

**Comment by woodsp-ibm** [comment]
A callback as alternative was discussed to a transpile engine (the latter I had always thought would have a method to be called that perhaps was the same as the callback but whose construction may take various parameters to set it up and possibly include setting backend after the fact too - potentially raising an error if ever used without a backend being set). An identity transpile engine is also possible allowing, where its possible with the algorithm, for a user to supply their own transpiled circuit - i.e. akin to skip_transpilation as the primitives had it before when they did transpilation.

Whether a global/default transpilation exists or not depends on what we want to happen about existing code. Clearly for runtime circuits need transpilation - can we have some default whereby it works no matter simulator or runtime. Clearly the code cannot work exactly as it stands, at least for runtime, since at least this engine would need the backend set so a suitable ISA circuit results. Hence perhaps a default, other than giving back identity, still would not get existing code to work - it still needs augmenting to define the backend target for transpilation however that manifests in the solution.

**Comment by 1ucian0** [comment]
For reference, an issue that it is consequence of this (with a workaround) here https://github.com/Qiskit/qiskit-ibm-runtime/issues/1718
